File: HavenApp/HavenApp/Services/ContactManager.swift
import Foundation

/// Pure contact list management logic — no Combine, no WebSocket.
/// FeedService delegates to these functions for contact list mutations
/// and validation. The WebSocket transport layer stays in FeedService.
enum ContactManager {

    /// Errors that can occur during follow/unfollow operations.
    enum FollowActionError: Error {
        case contactsNotLoaded
        case alreadyFollowing
        case cannotUnfollowSelf
    }

    // MARK: - Contact List Parsing

    /// Parses a kind-3 contact list event's p-tags into the canonical follow set.
    /// Ensures the owner and whitelisted accounts are included.
    ///
    /// - Parameters:
    ///   - pTags: Pre-filtered "p" tags from the kind-3 event (e.g. ["p", hex, relay?, petname?]).
    ///   - ownerHex: Hex pubkey of the current account owner.
    ///   - whitelistedNpubs: Npub strings of accounts to auto-follow.
    /// - Returns: Tuple of (finalPTags, followedPubkeys, relayPTagCount) where
    ///   relayPTagCount is the count from the relay before owner/whitelist additions.
    static func parseContactList(
        pTags: [[String]],
        ownerHex: String,
        whitelistedNpubs: [String]
    ) -> (pTags: [[String]], pubkeys: [String], relayPTagCount: Int) {
        let relayCount = pTags.count
        var finalPTags = pTags
        let existingPubkeys = Set(pTags.compactMap { $0.count >= 2 ? $0[1] : nil })

        // Ensure owner is in the list
        if !existingPubkeys.contains(ownerHex) {
            finalPTags.append(["p", ownerHex])
        }

        // Auto-follow whitelisted accounts
        for npub in whitelistedNpubs {
            if let hex = Bech32.decode(npub)?.hexString, !existingPubkeys.contains(hex) {
                finalPTags.append(["p", hex])
            }
        }

        let pubkeys = finalPTags.compactMap { $0.count >= 2 ? $0[1] : nil }
        return (finalPTags, pubkeys, relayCount)
    }

    // MARK: - Follow / Unfollow

    /// Validates and performs a follow operation on the tag list.
    /// Returns updated tags/pubkeys on success, or an error.
    static func prepareFollow(
        pubkey: String,
        currentPTags: [[String]],
        currentPubkeys: [String],
        hasAttemptedLoad: Bool,
        isLoading: Bool
    ) -> Result<(pTags: [[String]], pubkeys: [String]), FollowActionError> {
        guard hasAttemptedLoad, !isLoading else { return .failure(.contactsNotLoaded) }
        guard !currentPubkeys.contains(pubkey) else { return .failure(.alreadyFollowing) }
        var newPTags = currentPTags
        var newPubkeys = currentPubkeys
        newPTags.append(["p", pubkey])
        newPubkeys.append(pubkey)
        return .success((newPTags, newPubkeys))
    }

    /// Validates and performs an unfollow operation on the tag list.
    /// Returns updated tags/pubkeys on success, or an error.
    static func prepareUnfollow(
        pubkey: String,
        activeAccountHex: String,
        currentPTags: [[String]],
        currentPubkeys: [String],
        hasAttemptedLoad: Bool,
        isLoading: Bool
    ) -> Result<(pTags: [[String]], pubkeys: [String]), FollowActionError> {
        guard hasAttemptedLoad, !isLoading else { return .failure(.contactsNotLoaded) }
        guard pubkey != activeAccountHex else { return .failure(.cannotUnfollowSelf) }
        var newPTags = currentPTags
        var newPubkeys = currentPubkeys
        newPTags.removeAll { $0.count >= 2 && $0[1] == pubkey }
        newPubkeys.removeAll { $0 == pubkey }
        return .success((newPTags, newPubkeys))
    }

    // MARK: - Safety Checks

    /// Returns true if publishing the contact list should be BLOCKED because
    /// it would shrink drastically relative to the last relay-fetched count
    /// (prevents accidental follow-list wipes from stale data).
    static func shouldBlockPublish(currentTagCount: Int, lastFetchedCount: Int) -> Bool {
        guard lastFetchedCount > 10 else { return false }
        let ratio = Double(currentTagCount) / Double(lastFetchedCount)
        return ratio < 0.5
    }

    // MARK: - Extended Network

    /// Counts mutual follows from a single kind-3 event's tags, excluding
    /// the user's own follows. Used incrementally as events arrive from relays.
    static func countMutualFollows(
        eventTags: [[String]],
        excludeSet: Set<String>
    ) -> [String: Int] {
        var counts: [String: Int] = [:]
        for tag in eventTags {
            if tag.count >= 2, tag[0] == "p" {
                let pk = tag[1]
                if !excludeSet.contains(pk) {
                    counts[pk, default: 0] += 1
                }
            }
        }
        return counts
    }

    /// Ranks pubkeys by mutual-follow count and returns the top results.
    /// Call after all kind-3 events have been accumulated.
    static func rankExtendedNetwork(
        mutualCounts: [String: Int],
        maxResults: Int = 500
    ) -> [String] {
        let sorted = mutualCounts.sorted { $0.value > $1.value }
        return Array(sorted.prefix(maxResults).map { $0.key })
    }
}

File: NostrVault/app/src/main/java/com/nostrvault/service/ContactManager.kt
package com.nostrvault.service

/**
 * Port of ContactManager.swift -- pure contact list management logic.
 * No networking, no UI. Operates on tag lists and pubkey sets.
 */
object ContactManager {

    sealed class FollowActionError : Exception() {
        data object ContactsNotLoaded : FollowActionError()
        data object AlreadyFollowing : FollowActionError()
        data object CannotUnfollowSelf : FollowActionError()
    }

    data class ContactListResult(
        val pTags: List<List<String>>,
        val pubkeys: List<String>,
        val relayPTagCount: Int,
    )

    data class FollowResult(
        val pTags: List<List<String>>,
        val pubkeys: List<String>,
    )

    /**
     * Parse a kind-3 contact list event's p-tags into the canonical follow set.
     * Ensures the owner and whitelisted accounts are included.
     */
    fun parseContactList(
        pTags: List<List<String>>,
        ownerHex: String,
        whitelistedNpubs: List<String>,
    ): ContactListResult {
        val relayCount = pTags.size
        val finalPTags = pTags.toMutableList()
        val existingPubkeys = pTags.mapNotNull { if (it.size >= 2) it[1] else null }.toSet()

        // Ensure owner is in the list
        if (ownerHex !in existingPubkeys) {
            finalPTags.add(listOf("p", ownerHex))
        }

        // Auto-follow whitelisted accounts
        for (npub in whitelistedNpubs) {
            // TODO: bech32 decode npub to hex
            // For now, skip bech32 conversion -- will be added when Bech32 utility is ported
        }

        val pubkeys = finalPTags.mapNotNull { if (it.size >= 2) it[1] else null }
        return ContactListResult(finalPTags, pubkeys, relayCount)
    }

    /**
     * Validate and perform a follow operation on the tag list.
     */
    fun prepareFollow(
        pubkey: String,
        currentPTags: List<List<String>>,
        currentPubkeys: List<String>,
        hasAttemptedLoad: Boolean,
        isLoading: Boolean,
    ): Result<FollowResult> {
        if (!hasAttemptedLoad || isLoading) {
            return Result.failure(FollowActionError.ContactsNotLoaded)
        }
        if (pubkey in currentPubkeys) {
            return Result.failure(FollowActionError.AlreadyFollowing)
        }
        val newPTags = currentPTags + listOf(listOf("p", pubkey))
        val newPubkeys = currentPubkeys + pubkey
        return Result.success(FollowResult(newPTags, newPubkeys))
    }

    /**
     * Validate and perform an unfollow operation on the tag list.
     */
    fun prepareUnfollow(
        pubkey: String,
        activeAccountHex: String,
        currentPTags: List<List<String>>,
        currentPubkeys: List<String>,
        hasAttemptedLoad: Boolean,
        isLoading: Boolean,
    ): Result<FollowResult> {
        if (!hasAttemptedLoad || isLoading) {
            return Result.failure(FollowActionError.ContactsNotLoaded)
        }
        if (pubkey == activeAccountHex) {
            return Result.failure(FollowActionError.CannotUnfollowSelf)
        }
        val newPTags = currentPTags.filter { !(it.size >= 2 && it[1] == pubkey) }
        val newPubkeys = currentPubkeys.filter { it != pubkey }
        return Result.success(FollowResult(newPTags, newPubkeys))
    }

    /**
     * Returns true if publishing should be BLOCKED because the contact list
     * would shrink drastically (prevents accidental wipes from stale data).
     */
    fun shouldBlockPublish(currentTagCount: Int, lastFetchedCount: Int): Boolean {
        if (lastFetchedCount <= 10) return false
        val ratio = currentTagCount.toDouble() / lastFetchedCount.toDouble()
        return ratio < 0.5
    }

    /**
     * Count mutual follows from a kind-3 event's tags, excluding the user's own follows.
     */
    fun countMutualFollows(
        eventTags: List<List<String>>,
        excludeSet: Set<String>,
    ): Map<String, Int> {
        val counts = mutableMapOf<String, Int>()
        for (tag in eventTags) {
            if (tag.size >= 2 && tag[0] == "p") {
                val pk = tag[1]
                if (pk !in excludeSet) {
                    counts[pk] = (counts[pk] ?: 0) + 1
                }
            }
        }
        return counts
    }

    /**
     * Rank pubkeys by mutual-follow count and return the top results.
     */
    fun rankExtendedNetwork(
        mutualCounts: Map<String, Int>,
        maxResults: Int = 500,
    ): List<String> =
        mutualCounts.entries
            .sortedByDescending { it.value }
            .take(maxResults)
            .map { it.key }
}
